File: 트리/leetcode662(Medium)-Maximum Width of Binary Tree.py
# BFS : 36ms(98%)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        result = 0
        q = collections.deque([(root, 0)])
        while q:
            last_idx = q[-1][1]
            first_idx = q[0][1]
            result = max(result, last_idx - first_idx + 1)
            
            for _ in range(len(q)):
                node, idx = q.popleft()
                sub_idx = idx - first_idx
                if node.left:
                    q.append((node.left, sub_idx * 2 + 1))
                if node.right:
                    q.append((node.right, sub_idx * 2 + 2))
        
        return result

File: 트리/leetcode662(Medium)-Maximum Width of Binary Tree.js
// BFS : 82ms(77.40%)
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var widthOfBinaryTree = function (root) {
  if (!root) {
    return 0
  }

  let result = 0
  const q = [[root, 0]]

  while (q.length > 0) {
    const lastIdx = q[q.length - 1][1]
    const firstIdx = q[0][1]

    result = Math.max(result, lastIdx - firstIdx + 1)

    let len = q.length
    while (len--) {
      const [node, idx] = q.shift()
      const subIdx = idx - firstIdx
      if (node.left) q.push([node.left, subIdx * 2 + 1])
      if (node.right) q.push([node.right, subIdx * 2 + 2])
    }
  }

  return result
}
